test/jdk/java/nio/channels/DatagramChannel/SRTest.java

```diff
@@ -24,184 +24,220 @@
+                throw new RuntimeException("NioWriter threw exception: " + ex);
+            } finally {
+                log.println("NioWriter Finished");
             }
         }
+
+        @Override
+        public void close() throws IOException {
+            dc.close();
+        }
     }
 
-    public static class ClassicReader implements Sprintable {
+    public static class ClassicReader implements Runnable, AutoCloseable {
         final DatagramSocket ds;
-        Exception e = null;
 
         ClassicReader() throws IOException {
-            this.ds = new DatagramSocket();
+            InetSocketAddress address = new InetSocketAddress(InetAddress.getLoopbackAddress(), 0);
+            this.ds = new DatagramSocket(address);
         }
 
         int port() {
             return ds.getLocalPort();
         }
 
-        public void throwException() throws Exception {
-            if (e != null)
-                throw e;
-        }
-
         public void run() {
             try {
                 byte[] buf = new byte[256];
                 DatagramPacket dp = new DatagramPacket(buf, buf.length);
                 ds.receive(dp);
                 String received = new String(dp.getData());
-                log.println(received);
-                ds.close();
+                log.println("ClassicReader received: " + received);
             } catch (Exception ex) {
-                e = ex;
+                log.println("ClassicWriter [" + ds.getLocalAddress() +"]");
+                throw new RuntimeException("ClassicReader threw exception: " + ex);
+            } finally {
+                log.println("ClassicReader finished");
             }
         }
+
+        @Override
+        public void close() throws IOException {
+            ds.close();
+        }
     }
 
-    public static class NioReader implements Sprintable {
+    public static class NioReader implements Runnable, AutoCloseable {
         final DatagramChannel dc;
-        Exception e = null;
 
         NioReader() throws IOException {
-            this.dc = DatagramChannel.open().bind(new InetSocketAddress(0));
+            InetSocketAddress address = new InetSocketAddress(InetAddress.getLoopbackAddress(), 0);
+            this.dc = DatagramChannel.open().bind(address);
         }
 
         int port() {
             return dc.socket().getLocalPort();
         }
 
-        public void throwException() throws Exception {
-            if (e != null)
-                throw e;
-        }
-
         public void run() {
             try {
                 ByteBuffer bb = ByteBuffer.allocateDirect(100);
                 SocketAddress sa = dc.receive(bb);
                 bb.flip();
-                CharBuffer cb = Charset.forName("US-ASCII").
-                    newDecoder().decode(bb);
-                log.println("From: "+sa+ " said " +cb);
-                dc.close();
+                CharBuffer cb = StandardCharsets.US_ASCII.newDecoder().decode(bb);
```

Comment:
If a charset is used, then it should be used consistently - in both classic and nio readers and writers.
That is, whenever `String::getBytes` is called or `new String(byte[])` is called.

Reply:
Makes sense, I think I will opt for using `US_ASCII` across all of the classic and nio readers/writers.

Reply:
Using an 8-bit charset might be a better choice, US_ASCII is only 7bit and might produce some unmappable characters.
ISO_8859_1 would be a better choice.

Reply:
Fair enough. Is there any reason then why UTF-8 would be used over ISO_8859_1 then, given what you said above about character mapping?

Reply:
The message sent by the writer is "hello" - it only contains ASCII characters, so any of US-ASCII, UTF-8, or ISO-8859-1, which are all compatible with ASCII (meaning: all these encode ASCII chars in the same way) would work.
In fact any charset in which "hello" is mappable would work. We simply need to make sure to use the *same charset* when encoding or decoding.

Reply:
Its case by case, if you don't know how the stream of characters was encoded into a byte stream then you can't reliably decode it back to characters.  For SRTest, it doesn't matter, all the strings I see are ascii.
But note that String.getBytes() uses a platform dependent encoding. 
Until the platform encoding is UTF-8, its recommended to use String.getBytes(charset).
A lot of times UTF-8 is a good guess for operations that convert strings to bytes and back.

Reply:
Alright then so I will continue to to just use US_ASCII in this test as "hello" is mappable with it. Encoding and Decoding the string will be done with US_ASCII. Appreciate the feedback guys, thanks!

Reply:
Thanks Conor. Yes - we could work with byte arrays but strings are easier to print & log & more human friendly...

Reply:
ok,  but please use getBytes(StandardCharsets.US_ASCII) to get the bytes avoiding ambiguity.

Reply:
The most recent commit performs a static import of the charset with `import static java.nio.charset.StandardCharsets.US_ASCII;`. I feel that it is clear enough and keeps references to the Charset concise.